src/Identity/IdentityServer/BaseRequestValidator.cs

```diff
@@ -348,32 +348,13 @@ private async Task<bool> IsValidAuthTypeAsync(User user, string grantType)
             return true;
         }
 
-        // Is user apart of any orgs? Use cache for initial checks.
-        var orgs = (await _currentContext.OrganizationMembershipAsync(_organizationUserRepository, user.Id))
-            .ToList();
-        if (orgs.Any())
+        // Check if user belongs to any organization with an active SSO policy 
+        // If 'EnforceSsoPolicyForAllUsers' is set to true then SSO policy applies to all user types otherwise it does not apply to Owner or Admin
+        var minSsoPolicyUserType = _globalSettings.Sso.EnforceSsoPolicyForAllUsers ? OrganizationUserType.Owner : OrganizationUserType.User;
+        var ssoPoliciesApplicableToUser = await _policyService.GetPoliciesApplicableToUserAsync(user.Id, PolicyType.RequireSso, minSsoPolicyUserType, OrganizationUserStatusType.Confirmed);
+        if (ssoPoliciesApplicableToUser.Any())
         {
-            // Get all org abilities
-            var orgAbilities = await _applicationCacheService.GetOrganizationAbilitiesAsync();
```

Comment:
I'm a little unsure about getting rid of this "gate", we can read faster from the cache than we can from the db, so it might still be useful for performance reasons - i.e. it might avoid calling the `PolicyService` method if it's not required. On the other hand, peeking under the hood of `_currentContext.OrganizationMembershipAsync`, that involves a db call anyway, so I'm not sure if there's any benefit here. Any ideas?

Ping @justindbaur because I think you've done IdentityServer work before, any view on replacing this series of conditionals with just a single db call, and without using the `OrganizationAbilities` cache?

Reply:
(I do love how much it cleans up this code though!)

Reply:
We discussed this in Slack and Justin OK'd it